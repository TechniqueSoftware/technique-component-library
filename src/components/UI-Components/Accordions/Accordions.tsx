// @ts-nocheck
import * as React from 'react';

import { default as MUIAccordion } from '@material-ui/core/Accordion';
import AccordionSummary from '@material-ui/core/AccordionSummary';
import AccordionDetails from '@material-ui/core/AccordionDetails';

import Typography from '@material-ui/core/Typography';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import FormHelperText from '@material-ui/core/FormHelperText';

import { Theme } from '@material-ui/core';
import { makeStyles, createStyles } from '@material-ui/styles';

export interface Accordion {
  // Status
  defaultExpanded?: boolean;  // Default: false
  disabled?: boolean;         // Default: true
  // Summary
  title?: React.ReactNode | string;
  subTitle?: string;
  helperText?: string;
  rightSummary?: React.ReactNode | string;
  summary?: React.ReactNode | string;   // Instead of the above props, a custom summary would be applied
  // Details
  details: React.ReactNode | string;
}

export interface AccordionsProps {
  className?: string;
  panels: Accordion[];
  mode?:
    'free' |      // Applied by the user
    'accordion'   // Only one open at a time
  ;
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    title: {
      fontSize: theme.typography.pxToRem(15),
      fontWeight: theme.typography.fontWeightRegular,
    },
    subTitle: {
      fontSize: theme.typography.pxToRem(12),
      fontWeight: theme.typography.fontWeightBold,
    },
    summaryContainer: {
      display: 'flex',
      width: '100%',
    },
    summaryLeft: {
      flex: '1 1 100%',
    },
    summaryRight: {
      flex: '1 1 0',
      marginLeft: theme.spacing(3),
    },
  }),
);

const Accordions = (props: AccordionsProps) => {
  const classes = useStyles();
  const {
    className,
    panels,
    mode = 'free',
  } = props;

  const internalId = React.useRef<string>();
  if (!internalId.current) internalId.current = Math.random().toString().substr(2);
  const { current: id } = internalId;

  const [expandedPanels, setExpandedPanels] = React.useState<boolean[]>(panels.map(panel => !!panel.defaultExpanded));

  React.useEffect(() => {
    if (mode === 'accordion') {
      const firstExpanded = expandedPanels.findIndex(Boolean);
      if (firstExpanded === -1) setExpandedPanels(panels.map(() => false));
      setExpandedPanels(panels.map((v, index) => index === firstExpanded));
    }
  }, [mode]);

  const handleExpanded =
    (expandedIndex: number) =>
      (event: React.ChangeEvent, expanded: boolean) => {
        const newExpandedPanels = expandedPanels.concat();
        switch (mode) {
          case 'free':
            newExpandedPanels[expandedIndex] = expanded;
            break;
          case 'accordion':
            panels.forEach((p, index) => newExpandedPanels[index] = expanded && index === expandedIndex);
            break;
        }
        setExpandedPanels(newExpandedPanels);
      };

  return (
    <div className={className}>
      {panels.map(
        ({
           title,
           subTitle,
           helperText,
           rightSummary,
           summary = null,
           details,
           defaultExpanded,
           disabled,
         },
         index,
        ) => (
          <MUIAccordion
            key={index}
            defaultExpanded={defaultExpanded}
            expanded={expandedPanels[index]}
            disabled={disabled}
            onChange={handleExpanded(index)}
          >
            <AccordionSummary
              expandIcon={<ExpandMoreIcon/>}
              aria-controls={`panel-${index}-content-${id}`}
              id={`panel-${index}-header-${id}`}
            >
              <div className={classes.summaryContainer}>
                <div className={classes.summaryLeft}>
                  <Typography className={classes.title}>{title}</Typography>
                  {!!subTitle && <Typography className={classes.subTitle}>{subTitle}</Typography>}
                  {!!helperText && <FormHelperText>{helperText}</FormHelperText>}
                  {summary}
                </div>
                {!!rightSummary && <div className={classes.summaryRight}>
                  {rightSummary}
                </div>}
              </div>
            </AccordionSummary>
            <AccordionDetails>
              {details}
            </AccordionDetails>
          </MUIAccordion>
        )
      )}
    </div>
  );
};

export default Accordions;
