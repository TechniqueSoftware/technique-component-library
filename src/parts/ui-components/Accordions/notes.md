# Expansion Panels
This component wraps Material-UI's [Accordions](https://material-ui.com/components/accordion/) component. There is currently
no customization done. 

## JIRA
Related jira ticket: https://clubos.atlassian.net/browse/CR-784
