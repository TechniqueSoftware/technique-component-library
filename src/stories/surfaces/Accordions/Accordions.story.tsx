import * as React from 'react';
import { createCUICStoriesOf } from '../../utils/utils';
import { MODULE_NAME } from '../../constants';
import { select } from '@storybook/addon-knobs';

import { default as Accordions, Accordion } from '@clubos-ui-components/core/dist/components/Accordions/Accordions';

import clubOsTheme from '@clubos-ui-components/core/dist/themes/clubOS';
import { MuiThemeProvider } from '@material-ui/core';

const AccordionsStory = () => {

  const panels: Accordion[] = [
    {
      title: 'Title',
      subTitle: 'Subtitle',
      helperText: 'Helper text. (default: expanded)',
      defaultExpanded: true,
      rightSummary: 'Right Summary panel',
      details: 'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec pede justo, fringilla vel, aliquet nec, vulputate.',
    },
    {
      title: 'Seven hills',
      subTitle: 'Brewery',
      details: 'Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec pede justo, fringilla vel, aliquet nec, vulputate.',
    },
    {
      title: 'Fanta panda',
      subTitle: 'Chinese',
      helperText: 'Disabled',
      disabled: true,
      details: 'Parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec pede justo, fringilla vel, aliquet nec, vulputate.',
    },
    {
      title: 'Scottish Notice',
      subTitle: 'Pub',
      details: 'Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore veritatis et quasi architecto beatae vitae dicta sunt explicabo. Nemo enim ipsam voluptatem quia voluptas sit aspernatur aut odit aut fugit, sed quia consequuntur magni dolores eos qui.',
    },
    {
      title: '7stern',
      subTitle: 'Brewery',
      helperText: 'Map info included',
      rightSummary: <div>Phone 699111234123</div>,
      details: 'Li Europan lingues es membres del sam familie. Lor separat existentie es un myth. Por scientie, musica, sport etc, litot Europa usa li sam vocabular. Li lingues differe solmen in li grammatica, li pronunciation e li plu commun vocabules. Omnicos directe al desirabilite de un nov lingua franca: On refusa continuar payar custosi traductores. At solmen va esser necessi far uniform grammatica, pronunciation e plu sommun paroles. Ma quande lingues coalesce, li grammatica del resultant lingue es plu simplic e regulari quam ti del coalescent lingues. Li nov lingua franca va esser plu simplic e regulari quam li existent Europan lingues. It va esser tam simplic quam Occidental in fact, it va esser Occidental. A un Angleso it va semblar un simplificat Angles, quam un skeptic Cambridge amico dit me que Occidental es.Li Europan lingues es membres del sam familie. Lor separat existentie es un myth. Por scientie, musica, sport etc, litot Europa usa li sam vocabular. Li lingues differe solmen in li grammatica, li pronunciation e li plu commun vocabules. Omnicos directe al desirabilite de un nov lingua franca: On refusa continuar payar custosi traductores. At solmen va esser necessi far uniform grammatica, pronunciation e plu sommun paroles.',
    },
    {
      title: '1234567890',
      summary: (
        <>
          <h4>Hollandia</h4>
          <h5>An irish pub</h5>
          <p>Custom Summary</p>
        </>
      ),
      details: 'Li Europan lingues es membres del sam familie. Lor separat existentie es un myth. Por scientie, musica, sport etc, litot Europa usa li sam vocabular. Li lingues differe solmen in li grammatica, li pronunciation e li plu commun vocabules. Omnicos directe al desirabilite de un nov lingua franca: On refusa continuar payar custosi traductores. At solmen va esser necessi far uniform grammatica, pronunciation e plu sommun paroles. Ma quande lingues coalesce, li grammatica del resultant lingue es plu simplic e regulari quam ti del coalescent lingues. Li nov lingua franca va esser plu simplic e regulari quam li existent Europan lingues. It va esser tam simplic quam Occidental in fact, it va esser Occidental. A un Angleso it va semblar un simplificat Angles, quam un skeptic Cambridge amico dit me que Occidental es.Li Europan lingues es membres del sam familie. Lor separat existentie es un myth. Por scientie, musica, sport etc, litot Europa usa li sam vocabular. Li lingues differe solmen in li grammatica, li pronunciation e li plu commun vocabules. Omnicos directe al desirabilite de un nov lingua franca: On refusa continuar payar custosi traductores. At solmen va esser necessi far uniform grammatica, pronunciation e plu sommun paroles.',
    },
  ];

  const mode = select('Mode', ['free', 'accordion'], 'free');

  return (
    <MuiThemeProvider theme={clubOsTheme}>
      <Accordions
        mode={mode}
        panels={panels}
      />

    </MuiThemeProvider>
  );
};

createCUICStoriesOf(MODULE_NAME.SURFACES, module)
  .addCUICStory({
    render: () => <AccordionsStory/>,
    name: 'Expansion Panels',
    notes: require('./notes.md').default
  });
