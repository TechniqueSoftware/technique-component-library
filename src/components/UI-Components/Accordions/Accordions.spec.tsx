import Accordions from './Accordions';
import { shallow } from 'enzyme';
import * as React from 'react';

import clubOsTheme from '../../../themes/clubOS';
import { MuiThemeProvider } from '@material-ui/core/styles';

describe('Accordions', () => {

  test('should render with base props', () => {
    const shallowWrapper = shallow(
      <MuiThemeProvider theme={clubOsTheme}>
        <Accordions
          panels={[
            {
              title: 'Title',
              details: 'Lorem ipsum.',
            },
          ]}
        />
      </MuiThemeProvider>
    );
    expect(shallowWrapper.debug()).toMatchSnapshot();
  });

  test('should render with all props', () => {
    const shallowWrapper = shallow(
      <MuiThemeProvider theme={clubOsTheme}>
        <Accordions
          className="class-name"
          mode="accordion"
          panels={[
            {
              title: 'Title',
              subTitle: 'Subtitle',
              helperText: 'Helper text. (default: expanded)',
              defaultExpanded: true,
              rightSummary: 'Right Summary panel',
              details: 'Lorem ipsum.',
            },
            {
              summary: (
                <>
                  <h4>Hollandia</h4>
                </>
              ),
              details: 'Li Europan lingues es membres del sam familie. Lor separat existentie es un myth. Por scientie, musica, sport etc, litot Europa usa li sam vocabular. Li lingues differe solmen in li grammatica, li pronunciation e li plu commun vocabules. Omnicos directe al desirabilite de un nov lingua franca: On refusa continuar payar custosi traductores. At solmen va esser necessi far uniform grammatica, pronunciation e plu sommun paroles. Ma quande lingues coalesce, li grammatica del resultant lingue es plu simplic e regulari quam ti del coalescent lingues. Li nov lingua franca va esser plu simplic e regulari quam li existent Europan lingues. It va esser tam simplic quam Occidental in fact, it va esser Occidental. A un Angleso it va semblar un simplificat Angles, quam un skeptic Cambridge amico dit me que Occidental es.Li Europan lingues es membres del sam familie. Lor separat existentie es un myth. Por scientie, musica, sport etc, litot Europa usa li sam vocabular. Li lingues differe solmen in li grammatica, li pronunciation e li plu commun vocabules. Omnicos directe al desirabilite de un nov lingua franca: On refusa continuar payar custosi traductores. At solmen va esser necessi far uniform grammatica, pronunciation e plu sommun paroles.',
            },
          ]}
        />
      </MuiThemeProvider>
    );
    expect(shallowWrapper.debug()).toMatchSnapshot();
  });

});
