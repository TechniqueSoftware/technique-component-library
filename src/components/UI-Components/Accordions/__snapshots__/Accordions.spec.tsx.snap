// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordions should render with all props 1`] = `
"<ContextProvider value={{...}}>
  <Accordions className=\\"class-name\\" mode=\\"accordion\\" panels={{...}} />
</ContextProvider>"
`;

exports[`Accordions should render with base props 1`] = `
"<ContextProvider value={{...}}>
  <Accordions panels={{...}} />
</ContextProvider>"
`;
